lib/activerecord-bitemporal/visualizer.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+module ActiveRecord::Bitemporal
+  module Visualizer
+    # Figure is a two-dimensional array holding plotted lines and columns
+    class Figure < Array
+      def print(str, line: 0, column: 0)
+        self[line] ||= []
+        str.each_char.with_index(column) do |c, i|
+          self[line][i] = c
+        end
+      end
+
+      def to_s
+        map { |l| l&.map { |c| c || ' ' }&.join }.join("\n")
+      end
+    end
+
+    module_function
+
+    def visualize(record, height: 10, width: 40, highlight: true)
+      histories = record.class.ignore_valid_datetime.within_deleted.bitemporal_for(record)
+  
+      valid_times = (histories.map(&:valid_from) + histories.map(&:valid_to)).sort.uniq
+      transaction_times = (histories.map(&:transaction_from) + histories.map(&:transaction_to)).sort.uniq
```

Comment:
If the `histories` are already loaded, `ActiveRecord::Calculations#pluck` is just the same as `map(&:valid_from)`.
https://github.com/rails/rails/blob/v7.0.3/activerecord/lib/active_record/relation/calculations.rb#L194
https://github.com/rails/rails/blob/v7.0.3/activesupport/lib/active_support/core_ext/enumerable.rb#L201

I also believe that `map(&:valid_from)` is better in performance because it is much more efficient in this case to iterate the array than to execute SQL.

Reply:
Okey, Thanks :)